### ISSUE TITLE
Inaccuracy in charging for storage operations in `VMLogic`

### ISSUE THREAD

**Issue by Longarithm** [near/nearcore#6225]
In `storage_read`, `storage_write` and `storage_remove` we work with key-value pairs using `ValuePtr` abstraction.

At first we descend into desired trie node using `storage_get` and get hash of the value stored in `RuntimeExtValuePtr`. Then we call `Self::deref_value` which internally make a DB query using value hash and retrieves the actual value.

The inaccuracy is that in `storage_write` and `storage_remove` we pay for touching trie nodes ***before*** dereferencing value; in `storage_read` we do it ***after*** dereferencing (see e.g. [https://github.com/near/nearcore/blob/8c172826220580fa477e5de036dce21b1fe8fa63/runtime/near-vm-logic/src/logic.rs#L2286-L2288](https://github.com/near/nearcore/blob/8c172826220580fa477e5de036dce21b1fe8fa63/runtime/near-vm-logic/src/logic.rs#L2286-L2288)). But `Self::deref_value` also increments touching trie node counter, so in the first two cases we charge the cost one extra time.

I suggest to unify implementation and don't charge for touching node because we already have `storage_*` fees, so it will be less confusing. This will have some reduction effect - e.g. if we do it for this receipt [https://explorer.near.org/transactions/Hu2KEB3jYQYDsNXBq6MSAmVpJNjfTwF2K2NTcTisCiLa#5WFYQK87fhYG5ChgoA3UaVyVQEzrfxX3X6ypy9FFqUgd](https://explorer.near.org/transactions/Hu2KEB3jYQYDsNXBq6MSAmVpJNjfTwF2K2NTcTisCiLa#5WFYQK87fhYG5ChgoA3UaVyVQEzrfxX3X6ypy9FFqUgd) - we can save 16 Ggas (current cost) * 36 (number of writes) ~= 0.6 Tgas in comparison with 93 Tgas in total.

cc @mm-near @bowenwang1996 @matklad

**Comment by matklad** [comment]
Yeah, because `deref_value` is guaranteed to touch at most one node, and because we have a separate cost for "everything except trie traversal itself", I agree that it's valid to not charge here (or, on the contrary, to charge on every path). 

I am not sure it's worth doing though -- it won't make simpler, as we'd have to keep two versions anyway to maintain protocol compatibility. Ie, the amount of confusion would just increase.

**Comment by bowenwang1996** [comment]
> I am not sure it's worth doing though -- it won't make simpler, as we'd have to keep two versions anyway to maintain protocol compatibility. Ie, the amount of confusion would just increase.

It still helps us save some gas, but agree that it is not high priority

**Comment by matklad** [comment]
If the goal here is to save the gas, a simpler (in terms of implementation) approach would be to reduce base cost by one touching trie node cost (cc @jakmeier )

**Comment by jakmeier** [comment]
Yeah, decreasing the base read cost as compensation seems viable to me. It might be a bit confusing for anyone auditing or otherwise working with the code but with some well-placed comments I think it wouldn't be too bad.
